craft_store/publisher/_request.py

```diff
@@ -48,3 +49,114 @@ class ReleaseRequest(TypedDict):
     """A list of resources to attach to this release."""
     revision: int | None
     """The revision to release to the channel."""
+
+
+class MacaroonRequest(BaseModel):
+    """Request for issuing a macaroon."""
+
+    permissions: list[str] = Field(description="List of permissions to grant")
+    description: str = Field(description="Description of the macaroon usage")
+    ttl: int = Field(description="Time to live in seconds")
+    packages: list[dict[str, str]] | None = Field(
+        default=None, description="Package restrictions"
+    )
```

Comment:
It seems strange to make the parent object a pydantic model but not represent the model here. The dictionary here is something like:

```python
class IdPackage(TypedDict):
    id: str
    type: PackageType  # An enum of possible packages

class NamePackage(TypedDict):
    name: str
    type: PackageType

PackageDict = IdPackage | NamePackage
```